File: src/crypto/sodium_interop.cpp
#include "ecliptix/crypto/sodium_interop.hpp"
#include "ecliptix/crypto/sodium_secure_memory_handle.hpp"

namespace ecliptix::protocol::crypto {
    Result<Unit, SodiumFailure> SodiumInterop::Initialize() {
        std::call_once(init_flag_, [] {
            if (sodium_init() < 0) {
                initialized_.store(false, std::memory_order_release);
            } else {
                initialized_.store(true, std::memory_order_release);
            }
        });
        if (!initialized_.load(std::memory_order_acquire)) {
            return Result<Unit, SodiumFailure>::Err(
                SodiumFailure::InitializationFailed(
                    std::string(ErrorMessages::SODIUM_INIT_FAILED)));
        }
        return Result<Unit, SodiumFailure>::Ok(unit);
    }

    bool SodiumInterop::IsInitialized() noexcept {
        return initialized_.load(std::memory_order_acquire);
    }

    Result<Unit, SodiumFailure> SodiumInterop::SecureWipe(const std::span<uint8_t> buffer) {
        if (!IsInitialized()) {
            return Result<Unit, SodiumFailure>::Err(
                SodiumFailure::InitializationFailed(
                    std::string(ErrorMessages::NOT_INITIALIZED)));
        }
        if (buffer.empty()) {
            return Result<Unit, SodiumFailure>::Ok(unit);
        }
        if (buffer.size() > MAX_BUFFER_SIZE) {
            return Result<Unit, SodiumFailure>::Err(
                SodiumFailure::BufferTooLarge(
                    "Buffer size " + std::to_string(buffer.size()) +
                    " exceeds maximum " + std::to_string(MAX_BUFFER_SIZE)));
        }
        if (buffer.size() <= Constants::SMALL_BUFFER_THRESHOLD) {
            return WipeSmallBuffer(buffer);
        }

        return WipeLargeBuffer(buffer);
    }

    Result<Unit, SodiumFailure> SodiumInterop::SecureWipe(std::span<const uint8_t> buffer) {
        return SecureWipe(std::span(
            const_cast<uint8_t *>(buffer.data()),
            buffer.size()));
    }

    Result<Unit, SodiumFailure> SodiumInterop::WipeSmallBuffer(std::span<uint8_t> buffer) {
        try {
            sodium_memzero(buffer.data(), buffer.size());
            return Result<Unit, SodiumFailure>::Ok(unit);
        } catch (const std::exception &ex) {
            return Result<Unit, SodiumFailure>::Err(
                SodiumFailure::SecureWipeFailed(
                    "Failed to wipe small buffer: " + std::string(ex.what())));
        }
    }

    Result<Unit, SodiumFailure> SodiumInterop::WipeLargeBuffer(std::span<uint8_t> buffer) {
        try {
            sodium_memzero(buffer.data(), buffer.size());
            return Result<Unit, SodiumFailure>::Ok(unit);
        } catch (const std::exception &ex) {
            return Result<Unit, SodiumFailure>::Err(
                SodiumFailure::SecureWipeFailed(
                    "Failed to wipe large buffer: " + std::string(ex.what())));
        }
    }

    Result<bool, SodiumFailure> SodiumInterop::ConstantTimeEquals(
        const std::span<const uint8_t> a,
        const std::span<const uint8_t> b) {
        if (a.size() != b.size()) {
            return Result<bool, SodiumFailure>::Ok(false);
        }
        if (a.empty()) {
            return Result<bool, SodiumFailure>::Ok(true);
        }
        try {
            const int result = sodium_memcmp(a.data(), b.data(), a.size());
            return Result<bool, SodiumFailure>::Ok(result == ComparisonConstants::EQUAL);
        } catch (const std::exception &ex) {
            return Result<bool, SodiumFailure>::Err(
                SodiumFailure::ComparisonFailed(
                    std::string(ErrorMessages::CONSTANT_TIME_COMPARISON_FAILED) +
                    ": " + ex.what()));
        }
    }

    Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >, ProtocolFailure>
    SodiumInterop::GenerateX25519KeyPair(std::string_view key_purpose) {
        try {
            auto sk_handle_result = SecureMemoryHandle::Allocate(
                kX25519PrivateKeyBytes);
            if (sk_handle_result.IsErr()) {
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                    ProtocolFailure>::Err(
                    ProtocolFailure::FromSodiumFailure(
                        sk_handle_result.UnwrapErr()));
            }
            SecureMemoryHandle sk_handle = std::move(sk_handle_result).Unwrap();
            std::vector<uint8_t> sk_bytes = GetRandomBytes(kX25519PrivateKeyBytes);
            auto write_result = sk_handle.Write(std::span<const uint8_t>(sk_bytes));
            SecureWipe(std::span(sk_bytes));
            if (write_result.IsErr()) {
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                    ProtocolFailure>::Err(
                    ProtocolFailure::FromSodiumFailure(
                        write_result.UnwrapErr()));
            }
            std::vector<uint8_t> temp_sk(kX25519PrivateKeyBytes);
            if (auto read_result = sk_handle.Read(std::span(temp_sk)); read_result.IsErr()) {
                SecureWipe(std::span(temp_sk));
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                    ProtocolFailure>::Err(
                    ProtocolFailure::FromSodiumFailure(
                        read_result.UnwrapErr()));
            }
            std::vector<uint8_t> pk_bytes(kX25519PublicKeyBytes);
            if (crypto_scalarmult_base(pk_bytes.data(), temp_sk.data()) != SodiumConstants::SUCCESS) {
                SecureWipe(std::span(temp_sk));
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                    ProtocolFailure>::Err(
                    ProtocolFailure::DeriveKey(
                        "Failed to derive " + std::string(key_purpose) + " public key"));
            }
            SecureWipe(std::span(temp_sk));
            if (pk_bytes.size() != kX25519PublicKeyBytes) {
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                    ProtocolFailure>::Err(
                    ProtocolFailure::DeriveKey(
                        "Derived " + std::string(key_purpose) + " public key has incorrect size"));
            }
            return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                ProtocolFailure>::Ok(
                std::make_pair(std::move(sk_handle), std::move(pk_bytes)));
        } catch (const std::exception &ex) {
            return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t> >,
                ProtocolFailure>::Err(
                ProtocolFailure::KeyGeneration(
                    "Unexpected error generating " + std::string(key_purpose) +
                    " key pair: " + ex.what()));
        }
    }

    Result<std::pair<SecureMemoryHandle, std::vector<uint8_t>>, ProtocolFailure>
    SodiumInterop::GenerateEd25519KeyPair() {
        try {
            auto sk_handle_result = SecureMemoryHandle::Allocate(kEd25519SecretKeyBytes);
            if (sk_handle_result.IsErr()) {
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t>>,
                    ProtocolFailure>::Err(
                    ProtocolFailure::FromSodiumFailure(sk_handle_result.UnwrapErr()));
            }
            SecureMemoryHandle sk_handle = std::move(sk_handle_result).Unwrap();
            std::vector<uint8_t> pk(kEd25519PublicKeyBytes);
            std::vector<uint8_t> temp_sk(kEd25519SecretKeyBytes);
            if (crypto_sign_keypair(pk.data(), temp_sk.data()) != SodiumConstants::SUCCESS) {
                SecureWipe(std::span(temp_sk));
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t>>,
                    ProtocolFailure>::Err(
                    ProtocolFailure::KeyGeneration("Failed to generate Ed25519 key pair"));
            }
            auto write_result = sk_handle.Write(std::span<const uint8_t>(temp_sk));
            SecureWipe(std::span(temp_sk));
            if (write_result.IsErr()) {
                return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t>>,
                    ProtocolFailure>::Err(
                    ProtocolFailure::FromSodiumFailure(write_result.UnwrapErr()));
            }
            return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t>>,
                ProtocolFailure>::Ok(
                std::make_pair(std::move(sk_handle), std::move(pk)));
        } catch (const std::exception &ex) {
            return Result<std::pair<SecureMemoryHandle, std::vector<uint8_t>>,
                ProtocolFailure>::Err(
                ProtocolFailure::KeyGeneration(
                    "Unexpected error generating Ed25519 key pair: " +
                    std::string(ex.what())));
        }
    }

    std::vector<uint8_t> SodiumInterop::GetRandomBytes(const size_t size) {
        std::vector<uint8_t> buffer(size);
        randombytes_buf(buffer.data(), size);
        return buffer;
    }

    uint32_t SodiumInterop::GenerateRandomUInt32(const bool ensure_non_zero) {
        uint32_t value;
        do {
            value = randombytes_uniform(UINT32_MAX);
        } while (ensure_non_zero && value == 0);
        return value;
    }

    void *SodiumInterop::AllocateSecure(const size_t size) noexcept {
        if (!IsInitialized()) {
            return nullptr;
        }
        return sodium_malloc(size);
    }

    void SodiumInterop::FreeSecure(void *ptr) noexcept {
        if (ptr != nullptr) {
            sodium_free(ptr);
        }
    }
}

File: nuget/shared/src/Sodium/SodiumInterop.cs
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Sodium;

#if ECLIPTIX_SERVER
namespace Ecliptix.Protocol.Server.Sodium;
#else
namespace Ecliptix.Protocol.Client.Sodium;
#endif

internal static partial class SodiumInterop
{
    private const string LIB_SODIUM = ProtocolSystemConstants.Libraries.LIB_SODIUM;

    private const int MAX_BUFFER_SIZE = 1_000_000_000;

    private static readonly Result<Unit, SodiumFailure> InitializationResult = InitializeSodium();

    public static bool IsInitialized => InitializationResult.IsOk;

    [LibraryImport(LIB_SODIUM, EntryPoint = "sodium_init")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    private static partial int sodium_init();

    [LibraryImport(LIB_SODIUM, EntryPoint = "sodium_malloc")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    internal static partial IntPtr sodium_malloc(nuint size);

    [LibraryImport(LIB_SODIUM, EntryPoint = "sodium_free")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    internal static partial void sodium_free(IntPtr ptr);

    [LibraryImport(LIB_SODIUM, EntryPoint = "sodium_memzero")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    private static partial void sodium_memzero(IntPtr ptr, nuint length);

    [LibraryImport(LIB_SODIUM, EntryPoint = "sodium_memcmp")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    private static partial int sodium_memcmp(byte[] b1, byte[] b2, nuint length);

    [LibraryImport(LIB_SODIUM, EntryPoint = "sodium_memcmp")]
    [UnmanagedCallConv(CallConvs = [typeof(CallConvCdecl)])]
    private static unsafe partial int sodium_memcmp(void* b1, void* b2, nuint length);

    private static Result<Unit, SodiumFailure> InitializeSodium()
    {
        return Result<Unit, SodiumFailure>.Try(
            () =>
            {
                int result = sodium_init();
                const int dllImportSuccess = ProtocolSystemConstants.Numeric.DLL_IMPORT_SUCCESS;
                if (result < dllImportSuccess)
                {
                    throw new InvalidOperationException(SodiumFailureMessages.SODIUM_INIT_FAILED);
                }
            },
            ex => ex switch
            {
                DllNotFoundException dllEx => SodiumFailure.LibraryNotFound(
                    string.Format(SodiumFailureMessages.LIBRARY_LOAD_FAILED, LIB_SODIUM), dllEx),
                InvalidOperationException opEx when opEx.Message.Contains(SodiumExceptionMessagePatterns
                        .SODIUM_INIT_PATTERN) =>
                    SodiumFailure.INITIALIZATION_FAILED(SodiumFailureMessages.INITIALIZATION_FAILED, opEx),
                _ => SodiumFailure.INITIALIZATION_FAILED(SodiumFailureMessages.UNEXPECTED_INIT_ERROR, ex)
            }
        );
    }

    public static void SecureWipe(byte[]? buffer)
    {
        if (!IsInitialized)
        {
            Result<Unit, SodiumFailure>.Err(
                SodiumFailure.INITIALIZATION_FAILED(SodiumFailureMessages.NOT_INITIALIZED));
            return;
        }

        Result<byte[], SodiumFailure>
            .FromValue(buffer, SodiumFailure.BUFFER_TOO_SMALL(SodiumFailureMessages.BUFFER_NULL))
            .Bind(nonNullBuffer => nonNullBuffer switch
            {
                { Length: 0 } => Result<Unit, SodiumFailure>.Ok(Unit.Value),
                _ => Result<byte[], SodiumFailure>.Validate(
                        nonNullBuffer,
                        buf => buf.Length <= MAX_BUFFER_SIZE,
                        SodiumFailure.BUFFER_TOO_LARGE(
                            string.Format(SodiumFailureMessages.BUFFER_TOO_LARGE, nonNullBuffer.Length, MAX_BUFFER_SIZE)))
                    .Bind(validBuffer => validBuffer.Length <= Constants.SMALL_BUFFER_THRESHOLD
                        ? WipeSmallBuffer(validBuffer)
                        : WipeLargeBuffer(validBuffer))
            });
    }

    public static Result<(SodiumSecureMemoryHandle skHandle, byte[] pk), EcliptixProtocolFailure> GenerateX25519KeyPair(
        string keyPurpose)
    {
        try
        {
            Result<SodiumSecureMemoryHandle, SodiumFailure> allocResult =
                SodiumSecureMemoryHandle.Allocate(Constants.X_25519_PRIVATE_KEY_SIZE);
            if (allocResult.IsErr)
            {
                return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>.Err(allocResult.UnwrapErr()
                    .ToEcliptixProtocolFailure());
            }

            SodiumSecureMemoryHandle? skHandle = allocResult.Unwrap();

            byte[] skBytes = SodiumCore.GetRandomBytes(Constants.X_25519_PRIVATE_KEY_SIZE);
            try
            {
                Result<Unit, SodiumFailure> writeResult = skHandle.Write(skBytes);
                if (writeResult.IsErr)
                {
                    skHandle.Dispose();
                    return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>.Err(writeResult.UnwrapErr()
                        .ToEcliptixProtocolFailure());
                }
            }
            finally
            {
                SecureWipe(skBytes);
            }

            byte[] tempPrivCopy = new byte[Constants.X_25519_PRIVATE_KEY_SIZE];
            try
            {
                Result<Unit, SodiumFailure> readResult = skHandle.Read(tempPrivCopy);
                if (readResult.IsErr)
                {
                    skHandle.Dispose();
                    return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>.Err(readResult.UnwrapErr()
                        .ToEcliptixProtocolFailure());
                }

                Result<byte[], EcliptixProtocolFailure> deriveResult = Result<byte[], EcliptixProtocolFailure>.Try(
                    () => ScalarMult.Base(tempPrivCopy),
                    ex => EcliptixProtocolFailure.DeriveKey($"Failed to derive {keyPurpose} public key.", ex));

                if (deriveResult.IsErr)
                {
                    skHandle.Dispose();
                    return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>
                        .Err(deriveResult.UnwrapErr());
                }

                byte[] pkBytes = deriveResult.Unwrap();

                if (pkBytes.Length != Constants.X_25519_PUBLIC_KEY_SIZE)
                {
                    skHandle.Dispose();
                    SecureWipe(pkBytes);
                    return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>.Err(
                        EcliptixProtocolFailure.DeriveKey($"Derived {keyPurpose} public key has incorrect size."));
                }

                return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>.Ok((skHandle, pkBytes));
            }
            finally
            {
                SecureWipe(tempPrivCopy);
            }
        }
        catch (Exception ex)
        {
            return Result<(SodiumSecureMemoryHandle, byte[]), EcliptixProtocolFailure>.Err(
                EcliptixProtocolFailure.KeyGeneration($"Unexpected error generating {keyPurpose} key pair.", ex));
        }
    }

    public static Result<bool, SodiumFailure> ConstantTimeEquals(ReadOnlySpan<byte> a, ReadOnlySpan<byte> b)
    {
        if (a.Length != b.Length)
        {
            return Result<bool, SodiumFailure>.Ok(false);
        }

        if (a.IsEmpty)
        {
            return Result<bool, SodiumFailure>.Ok(true);
        }

        try
        {
            unsafe
            {
                fixed (byte* ptrA = a)
                fixed (byte* ptrB = b)
                {
                    int result = sodium_memcmp(ptrA, ptrB, (nuint)a.Length);
                    return Result<bool, SodiumFailure>.Ok(result == ProtocolSystemConstants.Numeric.ZERO_VALUE);
                }
            }
        }
        catch (Exception ex)
        {
            return Result<bool, SodiumFailure>.Err(
                SodiumFailure.ComparisonFailed(ProtocolSystemConstants.ErrorMessages.LIB_SODIUM_CONSTANT_TIME_COMPARISON_FAILED, ex));
        }
    }

    private static Result<Unit, SodiumFailure> WipeSmallBuffer(byte[] buffer)
    {
        return Result<Unit, SodiumFailure>.Try(
            () => { Array.Clear(buffer, ProtocolSystemConstants.Numeric.ZERO_VALUE, buffer.Length); },
            ex =>
                SodiumFailure.SECURE_WIPE_FAILED(
                    string.Format(SodiumFailureMessages.SMALL_BUFFER_CLEAR_FAILED, buffer.Length), ex));
    }

    private static Result<Unit, SodiumFailure> WipeLargeBuffer(byte[] buffer)
    {
        GCHandle handle = default;
        return Result<Unit, SodiumFailure>.Try(
            () =>
            {
                handle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
                IntPtr ptr = handle.AddrOfPinnedObject();
                if (ptr == IntPtr.Zero && buffer.Length > ProtocolSystemConstants.Numeric.ZERO_VALUE)
                {
                    throw new InvalidOperationException(SodiumFailureMessages.ADDRESS_OF_PINNED_OBJECT_FAILED);
                }

                sodium_memzero(ptr, (UIntPtr)buffer.Length);
            },
            ex => ex switch
            {
                ArgumentException argEx => SodiumFailure.MemoryPinningFailed(
                    SodiumFailureMessages.PINNING_FAILED, argEx),
                OutOfMemoryException oomEx => SodiumFailure.MemoryPinningFailed(
                    SodiumFailureMessages.INSUFFICIENT_MEMORY, oomEx),
                InvalidOperationException opEx when opEx.Message.Contains(SodiumExceptionMessagePatterns
                        .ADDRESS_PINNED_OBJECT_PATTERN) =>
                    SodiumFailure.MemoryPinningFailed(SodiumFailureMessages.GET_PINNED_ADDRESS_FAILED, opEx),
                _ => SodiumFailure.MemoryPinningFailed(
                    string.Format(SodiumFailureMessages.SECURE_WIPE_FAILED, buffer.Length), ex)
            },
            () =>
            {
                if (handle.IsAllocated)
                {
                    handle.Free();
                }
            }
        );
    }
}
